src/mgr/TTLCache.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <Python.h>
+#include <chrono>
+#include <fstream>
+#include <map>
+#include <memory>
+#include <string>
+
+using namespace std;
+
+// for reference 1000 osds is around 4 mb
+constexpr auto MAX_CACHE_SIZE = 4096;
+
+template <class Value> class Cache {
+protected:
+  Cache(int size = MAX_CACHE_SIZE) : capacity{size}, current_size{0} {};
+
+public:
+  int capacity;
+  int current_size;
+  virtual void insert(string key, Value value) = 0;
+  virtual Value get(string key) = 0;
+  virtual void erase(string key) = 0;
+  virtual void clear() = 0;
+  virtual ~Cache(){};
+};
+
+template <class Value> class TTLCache : public Cache<Value> {
+private:
+  int ttl;
+  typedef std::chrono::time_point<std::chrono::system_clock> time_stored;
+  typedef std::pair<Value, time_stored> value_type;
+  std::map<string, std::pair<Value, time_stored>> content;
+
+  int get_size(string key);
+  Value get_value(string key);
+  bool exists(string key);
+
+public:
+  TTLCache(int ttl_, int max_size = MAX_CACHE_SIZE)
+      : Cache<Value>(max_size), ttl{ttl_} {
+  }
+  ~TTLCache(){};
+  void insert(string key, Value value);
+  Value get(string key);
+  void erase(string key);
+  void clear();
+  void set_ttl(int ttl);
+  int get_pyobj_size(PyObject *obj);
+};
+
+template <class Value> void TTLCache<Value>::insert(string key, Value value) {
+  auto now = std::chrono::system_clock::now();
+  if (this->current_size + size <= this->capacity) {
+    content.insert({key, {value, now}});
+    // TODO: update size
+  }
+}
+
+template <class Value> Value TTLCache<Value>::get(string key) {
+  if (!this->exists(key)) {
+    return nullptr;
+  }
+  value_type stored_value = content[key];
+  time_stored stored_value_time = std::get<1>(stored_value);
+  auto end = std::chrono::system_clock::now();
+  int elapsed_time =
+      chrono::duration_cast<chrono::milliseconds>(end - stored_value_time)
+          .count();
+  if (elapsed_time >= ttl * 1000) {
+    erase(key);
```

Comment:
As cache reads are more frequent than writes, you could better calculate the expiration timestamp when storing and just comparing that on each read. Additionally, as discussed, to mitigate the performance impact of all cached entries expiring at the same time, you could add some randomization (e.g.: 25%) around the expiration time.